include/openssl/store.h

```diff
@@ -120,11 +120,12 @@ int OSSL_STORE_close(OSSL_STORE_CTX *ctx);
  * Note that this function is considered unsafe, all depending on what the
  * BIO actually reads.
  */
-OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bio, OPENSSL_CTX *libctx,
-                                  const char *scheme, const char *propq,
+OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bio,
+                                  const char *scheme,
                                   const UI_METHOD *ui_method, void *ui_data,
                                   OSSL_STORE_post_process_info_fn post_process,
-                                  void *post_process_data);
+                                  void *post_process_data,
+                                  OPENSSL_CTX *libctx, const char *propq);
```

Comment:
The previous pattern was modelled around the typical fetch call, i.e. starting with the tuple \< libctx, name, propq \>.  Is there a reason to shift that around?  Should we remodel the fetch calls accordingly?

I believe the logic for the fetch call model is "fetch within the *library context* something called *name*, using the given *properties*"

Reply:
And yeah, I was expecting the new opening function to look like this:

``` C

OSSL_STORE_CTX *OSSL_STORE_open_with_libctx(
    OPENSSL_CTX *libctx, const char *uri, const char *propq,
    const UI_METHOD *ui_method, void *ui_data,
    OSSL_STORE_post_process_info_fn post_process, void *post_process_data);
```

Reply:
I would like to stuff them all on the end for all API's (plus find any remaining _ex ones)..
I don't think it changes the language of the fetch much.
i.e: "fetch something called name with the given library context and properties".
@paulidale also mentioned maybe having a struct that contains both the libctx and propq and passing that around instead (since it is almost always done as a pair).

I think @mspncp also mentioned putting the libctx on the end of the list also.
Fixing the existing ones up is another PR though - probably best left to just before beta (when hopefully big PR's are settling down).

Reply:
I have no problems re-arranging them into a \< name, libctx, propq \> tuple, but please keep them together at least.

Reply:
In other words, something like this then:

``` C

OSSL_STORE_CTX *OSSL_STORE_open_with_libctx(
    const char *uri, OPENSSL_CTX *libctx, const char *propq,
    const UI_METHOD *ui_method, void *ui_data,
    OSSL_STORE_post_process_info_fn post_process, void *post_process_data);
```

Reply:
(fetch functions can easily be remodelled like that)

Reply:
As for a struct, that invites quite clumsy programming for not much extra value.  I'm not so keen on that.

Reply:
It does until we need to add yet another parameter associated with a libctx and then end up with
_with_libctx_ex :)

Reply:
Unless you make the struct non-public with getters and setters which would be a really horrible to use by callers, it will not help for such kind of API break.

Reply:
Please change this signature to, so it's consistent with OSSL_STORE_open_with_libctx:

``` C
OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bio, const char *scheme,
                                  OPENSSL_CTX *libctx, const char *propq,
                                  const UI_METHOD *ui_method, void *ui_data,
                                  OSSL_STORE_post_process_info_fn post_process,
                                  void *post_process_data);
```